src/boot/efi/secure-boot.c

```diff
@@ -236,10 +244,39 @@ void install_security_override(security_validator_t validator, const void *valid
                         (void **) &security_override.original_security2);
 }
 
+/* Some firmware will happily replace their own security arch instance with our override using
+ * ReinstallProtocolInterface, but not end up using it. In those cases we instead hack into it by
+ * swaping the function pointer with our own. We only do this if we have to as the reinstall method
+ * is much cleaner. */
+bool install_security_hack(void) {
+        if (security_override.was_called)
+                return false;
```

Comment:
this deserves a comment, i.e. `/* only do this hack if we know for sure the cleaner solution didn't work and we weren't called at all */` or something like that.